types/wlr_output_management_v1.c

```diff
@@ -0,0 +1,860 @@
+
+	if (config_head != NULL) {
+		config_head->resource = resource;
+		config_head->state.enabled = true;
+	}
+}
+
+static void config_handle_disable_head(struct wl_client *client,
+		struct wl_resource *config_resource,
+		struct wl_resource *head_resource) {
+	struct wlr_output_configuration_v1 *config =
+		config_from_resource(config_resource);
+	if (config == NULL || config->finalized) {
+		wl_resource_post_error(config_resource,
+			ZWLR_OUTPUT_CONFIGURATION_V1_ERROR_ALREADY_USED,
+			"configuration object has already been used");
+		return;
+	}
+	struct wlr_output_head_v1 *head = head_from_resource(head_resource);
+	if (head == NULL) {
+		return;
+	}
+
+	if (!config_check_head_is_unconfigured(config, head->state.output)) {
+		return;
+	}
+	struct wlr_output_configuration_head_v1 *config_head =
+		config_head_create(config, head->state.output);
+	if (config_head == NULL) {
+		wl_resource_post_no_memory(config_resource);
+		return;
+	}
+
+	config_head->state.enabled = false;
+}
+
+static void config_finalize(struct wlr_output_configuration_v1 *config) {
+	if (config->finalized) {
+		return;
+	}
+
+	// Destroy config head resources now, the client is forbidden to use them at
+	// this point anyway
+	struct wlr_output_configuration_head_v1 *config_head, *tmp;
+	wl_list_for_each_safe(config_head, tmp, &config->heads, link) {
+		// Resource is NULL if head has been disabled
+		if (config_head->resource != NULL) {
+			wl_resource_set_user_data(config_head->resource, NULL);
+			wl_resource_destroy(config_head->resource);
+			config_head->resource = NULL;
+		}
+	}
+
+	config->finalized = true;
+}
+
+// Destroys the config if serial is invalid
+static bool config_validate_serial(struct wlr_output_configuration_v1 *config) {
+	if (config->serial != config->manager->serial) {
+		wlr_log(WLR_DEBUG, "Ignored configuration request: invalid serial");
+		zwlr_output_configuration_v1_send_cancelled(config->resource);
+		wlr_output_configuration_v1_destroy(config);
+		return false;
+	}
+	return true;
+}
+
+static void config_handle_apply(struct wl_client *client,
+		struct wl_resource *config_resource) {
+	struct wlr_output_configuration_v1 *config =
+		config_from_resource(config_resource);
+	if (config == NULL || config->finalized) {
+		wl_resource_post_error(config_resource,
+			ZWLR_OUTPUT_CONFIGURATION_V1_ERROR_ALREADY_USED,
+			"configuration object has already been used");
+		return;
+	}
+
+	config_finalize(config);
+	if (!config_validate_serial(config)) {
+		return;
+	}
+
+	wlr_signal_emit_safe(&config->manager->events.apply, config);
+}
+
+static void config_handle_test(struct wl_client *client,
+		struct wl_resource *config_resource) {
+	struct wlr_output_configuration_v1 *config =
+		config_from_resource(config_resource);
+	if (config == NULL || config->finalized) {
+		wl_resource_post_error(config_resource,
+			ZWLR_OUTPUT_CONFIGURATION_V1_ERROR_ALREADY_USED,
+			"configuration object has already been used");
```

Comment:
Hmm. We don't want to send a protocol error if a configuration object has been cancelled:

```
client sends test()
server sends cancelled()
server receives test()
# This is not a protocol error
```

I overlooked this, thanks for spotting it. If the config is inert, we'll want to do nothing.

Reply:
Oh, in fact, we only send the `cancelled` event after an `apply`/`test` request so this can't happen.

So this can only happen if the client sends multiple `apply`/`test` requests.